plugins/debuginfo-install.py

```diff
@@ -79,54 +83,95 @@ def run(self):
         errors_spec = []
 
         for pkgspec in self.opts.package:
-            package_query = sorted(dnf.subject.Subject(pkgspec).get_best_query(
-                self.cli.base.sack).filter(arch__neq='src'), reverse=True)
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
+
+            packages = query.available()._na_dict()
+            # installed versions of packages have priority, replace / add them to the dict
+            packages.update(query.installed()._na_dict())
+
+            # Remove debuginfo packages if their base packages are in the query.
+            # They can get there through globs and they break the installation
+            # of debug packages with the same version as the installed base
+            # packages. If the base package of a debuginfo package is not in
+            # the query, the user specified a debug package on the command
+            # line. We don't want to ignore those, so we will install them.
+            # But, in this case the version will not be matched to the
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version himself etc.
+            for na in list(packages.keys()):
```

Comment:
I would recommend to use a query instead with query = query.union(query2) or difference or so on functions.

Reply:
How exactly would you use it in this case? I am doing a check for each package I'm iterating through. An alternative could be to make a list of packages that match the condition and filter those out - but the list can be potentially very long, and it will be a list of (name, architecture) pairs. Is this what you mean?

Reply:
So, the root cause why you are doing it is that ``debuginfo-install foo-debuginfo`` installs the newest version and does not match the version of ``foo``. And in the specific case of ``debuginfo-install foo foo-debuginfo`` (or similarly with globs), this fixes the problem. Right?

I get that you cannot just naively check the name without the suffix... but you are actually doing that here. So why not just check the existence of the base package, not just its presence in the packages? Would it be too slow? I really don't know, so sorry if you already thought about this.

I am thinking about doing one query for all the base names, installing these and then just checking that all the debuginfo names are in the packages_installed; and if not, installing them, too. Also, I don't know what @j-mracek had in mind, so maybe he has some good idea.

Reply:
> So, the root cause why you are doing it is that `debuginfo-install foo-debuginfo` installs the newest version and does not match the version of `foo`. And in the specific case of `debuginfo-install foo foo-debuginfo` (or similarly with globs), this fixes the problem. Right?

Yes.

> I get that you cannot just naively check the name without the suffix... but you are actually doing that here. So why not just check the existence of the base package, not just its presence in the packages? Would it be too slow? I really don't know, so sorry if you already thought about this.
> 
> I am thinking about doing one query for all the base names, installing these and then just checking that all the debuginfo names are in the packages_installed; and if not, installing them, too. Also, I don't know what @j-mracek had in mind, so maybe he has some good idea.

I'm not sure how this would be better? I still need to separate the packages to "base" and "debuginfo" by checking for the suffix, and then I'd have to do this in two steps, which I think would also result in not installing a newer version of e.g. `foo-debuginfo` with `dnf debuginfo-install foo-debuginfo` if an older version of `foo-debuginfo` was already installed? But maybe I'm just not getting it...

If you are thinking of performance overhead of using the NA (name, architecture) dicts, I think it's not a concern, I need to use the dicts anyway (to squash the installed and available package lists) and then the lookup is just O(1). The number of packages also shouldn't be huge.